File: tools/convsym/src/output/DEB1.cpp

/* ------------------------------------------------------------ *
 * ConvSym utility version 2.12									*
 * Output wrapper for the Debug Information format version 1.0	*
 * ------------------------------------------------------------	*/

#include <cassert>
#include <map>
#include <cstdint>
#include <string>

#include <OptsParser.hpp>
#include <Huffman.hpp>
#include <BitStream.hpp>
#include <IO.hpp>
#include <utils.hpp>

#include "OutputWrapper.hpp"


struct Output__Deb1 : public OutputWrapper {

	Output__Deb1() {};
	~Output__Deb1() {};

	/**
	 * Main function that generates the output
	 */
	void parse(
		std::multimap<uint32_t, std::string>& symbols,
		const char * fileName,
		uint32_t appendOffset = 0,
		uint32_t pointerOffset = 0,
		const char * opts = "",
		bool alignOnAppend = true
	) {
		assert(!symbols.empty());

		auto output = OutputWrapper::setupOutput(fileName, appendOffset, pointerOffset, alignOnAppend);

		/* Parse options from "-inopt" agrument's value */
		bool optFavorLastLabels = false;

		const std::map<std::string, OptsParser::record>
			OptsList {
				{ "favorLastLabels",	{ .type = OptsParser::record::p_bool,	.target =&optFavorLastLabels } }
			};
			
		OptsParser::parse(opts, OptsList);

		/* Write format version token */
		output->writeBEWord(0xDEB1);

		/* Allocate space for blocks offsets table */
		auto lastSymbolPtr = symbols.rbegin();
		uint16_t lastBlock = (lastSymbolPtr->first) >> 16;

		if (lastBlock > 63) {		// blocks index table is limited to $40 entries (which is only enough to ROM section)
			IO::Log( IO::error, "Too many memory blocks to allocate ($%X), truncating to $40 blocks. Some symbols will be lost.", lastBlock+1 );
			lastBlock = 0x3F;
		}

		uint16_t blockOffsets[0x40] = { 0 };
		uint16_t dataOffsets[0x40] = { 0 };

		const uint32_t loc_BlockOffsets = output->getCurrentOffset();	// remember the offset where blocks offset table should start
		output->setOffset(sizeof(blockOffsets) + sizeof(dataOffsets), IO::current);	// reserve space to write down offsets tables later

		/* ------------------------------------------------ */
		/* Generate Huffman-codes and create decoding table */
		/* ------------------------------------------------ */

		IO::Log(IO::debug, "Building an encoding table...");

		/* Generate table of character frequencies based on symbol names */
		uint32_t freqTable[0x100] = { 0 };
		for (auto& symbol : symbols) {
			for (auto& character : symbol.second) {
				freqTable[(int)character]++;
			}
			freqTable[0x00]++;
		}

		/* Generate table of Huffman codes (sorted by code) */
		Huffman::RecordSet codesTable = Huffman::encode(freqTable);

		/* Write down the decoding table */
		const Huffman::Record* characterToRecord[0x100] = { nullptr };	// LUT that links each character to its Huffman-coding record
		for (auto& entry : codesTable) {
			characterToRecord[entry.data] = &entry;		// assign character this Huffman::Record entity
			output->writeBEWord(entry.code);			// write Huffman-code
			output->writeByte(entry.codeLength);		// write Huffman-code length (in bits)
			output->writeByte(entry.data);				// write down original character
		}
		output->writeWord(-1);			// write 0xFFFF at the end of Huffman-table to stop searching and cause error

		/* ------------------------------------- */
		/* Generate per block symbol information */
		/* ------------------------------------- */

		{
			IO::Log(IO::debug, "Generating symbol data blocks...");

			auto symbolPtr = symbols.begin();

			/* For 64kb block within symbols range */
			for (uint16_t block = 0x00; block <= lastBlock; block++) {

				/* Align block on even address */
				if (output->getCurrentOffset() & 1) {
					output->writeByte(0x00);
				}

				IO::Log( IO::debug, "Block %02X:", block);
				uint32_t loc_Block = output->getCurrentOffset();	// remember offset, where this block starts ...

				std::vector<uint16_t> offsetsData;
				std::vector<uint8_t> symbolsData;

				/* For every symbol within the block ... */
				for (; (symbolPtr->first>>16) <= block && (symbolPtr != symbols.cend()); ++symbolPtr) {
					if ((symbolPtr->first>>16) < block) {
						continue;
					}

					IO::Log( IO::debug, "\t%08X\t%s", symbolPtr->first, symbolPtr->second.c_str() );

					/* 
					 * For records with the same offsets, fetch only the last or the first processed symbol,
					 * depending "favor last labels" option ...
					 */
					if ( (optFavorLastLabels && std::next(symbolPtr) != symbols.end()
								&& std::next(symbolPtr)->first == symbolPtr->first) ||
						 (!optFavorLastLabels && symbolPtr != symbols.begin()
								&& std::prev(symbolPtr)->first == symbolPtr->first) ) {
						continue;
					}

					BitStream symbolHeap;

					/* Add offset to the offsets block */
					offsetsData.push_back(swap16((uint16_t)symbolPtr->first & 0xFFFF));

					/* Encode each symbol character with the generated Huffman-codes and store it in the bitsteam */
					for (auto& Character : symbolPtr->second) {
						auto *record = characterToRecord[(int)Character];
						symbolHeap.pushCode(record->code, record->codeLength);
					}

					/* Finally, add null-terminator */
					{
						auto *record = characterToRecord[0x00];
						symbolHeap.pushCode(record->code, record->codeLength);
					}

					/* Push this symbol to the data buffer */
					symbolsData.push_back(symbolHeap.size() + 1);	// write down symbols size
					for (auto t = symbolHeap.begin(); t != symbolHeap.end(); t++) {
						symbolsData.push_back(*t);
					}
				}

				/* Write offsets block and their corresponding encoded symbols heap */
				if (offsetsData.size() > 0) {
					/* Add zero offset to finalize the block */
					offsetsData.push_back(0x0000);
					
					/* Check for pointer capacity limits */
					if ( (loc_Block - loc_BlockOffsets)>>1 > 0xFFFF
						|| (loc_Block+offsetsData.size()*2 - loc_BlockOffsets)>>1 > 0xFFFF ) {
						IO::Log( IO::error, "Block %02X is either too large, or symbol file has exceeded its size limits; unable to write the block", block );
						continue;
					}

					/* Setup pointers to the blocks */
					blockOffsets[block] = swap16((loc_Block - loc_BlockOffsets)>>1);
					dataOffsets[block] = swap16((loc_Block+offsetsData.size()*2 - loc_BlockOffsets)>>1);

					/* Write down block offsets and symbol data */
					output->writeData(&offsetsData[0], offsetsData.size()*sizeof(uint16_t));
					output->writeData(&symbolsData[0], symbolsData.size());
				}
			}
		}
		
		/* Finally, write down block offsets table in the header */
		output->setOffset(loc_BlockOffsets, IO::start);
		output->writeData(blockOffsets, sizeof(blockOffsets));
		output->writeData(dataOffsets, sizeof(dataOffsets));
	};

};

File: tools/crates/convsym/src/output/deb1.rs
/// Output generator for Debug Information format version 1.0 (DEB1).
/// Uses Huffman encoding for symbol names.

use std::collections::HashMap;
use std::io::{Seek, SeekFrom, Write};

use anyhow::Result;

trait WriteSeek: Write + Seek {}
impl<T: Write + Seek> WriteSeek for T {}

use crate::bitstream::BitStream;
use crate::huffman;
use crate::opts_parser::{self, RecordTarget};
use crate::output::OutputGenerator;

pub struct Deb1Output;

/// Helper to write big-endian values to a writer.
fn write_be_word(w: &mut impl Write, val: u16) -> Result<()> {
    w.write_all(&val.to_be_bytes())?;
    Ok(())
}

fn write_word(w: &mut impl Write, val: u16) -> Result<()> {
    w.write_all(&val.to_le_bytes())?;
    Ok(())
}

impl OutputGenerator for Deb1Output {
    fn generate(
        &self,
        symbols: &[(u32, String)],
        file_name: &str,
        append_offset: i64,
        pointer_offset: u32,
        opts: &str,
        align_on_append: bool,
    ) -> Result<()> {
        if symbols.is_empty() {
            anyhow::bail!("No symbols to output");
        }

        let mut output = setup_output(file_name, append_offset, pointer_offset, align_on_append)?;

        // Options
        let mut favor_last_labels = false;
        {
            let mut records: HashMap<&str, RecordTarget> = HashMap::new();
            records.insert(
                "favorLastLabels",
                RecordTarget::Bool(&mut favor_last_labels),
            );
            opts_parser::parse(opts, &mut records);
        }

        // Write format version token
        write_be_word(&mut output.writer, 0xDEB1)?;
        output.pos += 2;

        // Calculate last block
        let last_symbol_offset = symbols.last().unwrap().0;
        let mut last_block = (last_symbol_offset >> 16) as u16;
        if last_block > 63 {
            eprintln!(
                "Error: Too many memory blocks (${:X}), truncating to $40 blocks",
                last_block + 1
            );
            last_block = 0x3F;
        }

        // Reserve space for block offsets tables
        let mut block_offsets = vec![0u16; 0x40];
        let mut data_offsets = vec![0u16; 0x40];

        let loc_block_offsets = output.current_offset();
        // Reserve space: 0x40 * 2 bytes for block_offsets + 0x40 * 2 bytes for data_offsets
        let reserved_size = 0x40 * 2 + 0x40 * 2;
        output.writer.write_all(&vec![0u8; reserved_size])?;
        output.pos += reserved_size;

        // Build frequency table from all symbol names
        let mut freq_table = [0u32; 256];
        for (_, label) in symbols {
            for &b in label.as_bytes() {
                freq_table[b as usize] += 1;
            }
            freq_table[0] += 1; // null terminator
        }

        // Generate Huffman codes
        let codes_table = huffman::encode(&freq_table, 16);

        // Build character-to-record LUT
        let mut char_to_record: [Option<(u32, u8)>; 256] = [None; 256];
        for record in &codes_table {
            char_to_record[record.data as usize] = Some((record.code, record.code_length));

            // Write Huffman table entry
            write_be_word(&mut output.writer, record.code as u16)?;
            output.writer.write_all(&[record.code_length])?;
            output.writer.write_all(&[record.data as u8])?;
            output.pos += 4;
        }
        // Write 0xFFFF terminator
        write_word(&mut output.writer, 0xFFFF)?;
        output.pos += 2;

        // Generate per-block symbol information
        let mut sym_idx = 0;

        for block in 0..=last_block {
            // Align on even address
            if output.current_offset() & 1 != 0 {
                output.writer.write_all(&[0])?;
                output.pos += 1;
            }

            let loc_block = output.current_offset();

            let mut offsets_data: Vec<u16> = Vec::new();
            let mut symbols_data: Vec<u8> = Vec::new();

            while sym_idx < symbols.len() {
                let (offset, ref label) = symbols[sym_idx];
                let sym_block = (offset >> 16) as u16;

                if sym_block > block {
                    break;
                }
                if sym_block < block {
                    sym_idx += 1;
                    continue;
                }

                // Handle duplicate offsets
                if should_skip_duplicate(symbols, sym_idx, favor_last_labels) {
                    sym_idx += 1;
                    continue;
                }

                let mut symbol_heap = BitStream::new();

                // Add offset to offsets block (big-endian)
                offsets_data.push((offset & 0xFFFF) as u16);

                // Encode symbol characters
                for &b in label.as_bytes() {
                    if let Some((code, code_len)) = char_to_record[b as usize] {
                        symbol_heap.push_code(code, code_len);
                    }
                }
                // Null terminator
                if let Some((code, code_len)) = char_to_record[0] {
                    symbol_heap.push_code(code, code_len);
                }

                // Push to data buffer
                symbols_data.push(symbol_heap.size() as u8 + 1);
                symbols_data.extend_from_slice(symbol_heap.data());

                sym_idx += 1;
            }

            if !offsets_data.is_empty() {
                // Add zero offset to finalize
                offsets_data.push(0x0000);

                // Check pointer capacity
                let block_rel = (loc_block - loc_block_offsets) >> 1;
                let data_rel = (loc_block + offsets_data.len() * 2 - loc_block_offsets) >> 1;
                if block_rel > 0xFFFF || data_rel > 0xFFFF {
                    eprintln!(
                        "Error: Block {:02X} exceeds size limits",
                        block
                    );
                    continue;
                }

                block_offsets[block as usize] = swap16(block_rel as u16);
                data_offsets[block as usize] = swap16(data_rel as u16);

                // Write offsets (big-endian)
                for off in &offsets_data {
                    write_be_word(&mut output.writer, *off)?;
                    output.pos += 2;
                }
                // Write symbol data
                output.writer.write_all(&symbols_data)?;
                output.pos += symbols_data.len();
            }
        }

        // Write block offsets table at the reserved position
        output.writer.seek(SeekFrom::Start(
            (output.base_offset + loc_block_offsets) as u64,
        ))?;
        for off in &block_offsets {
            output.writer.write_all(&off.to_ne_bytes())?;
        }
        for off in &data_offsets {
            output.writer.write_all(&off.to_ne_bytes())?;
        }

        Ok(())
    }
}

fn swap16(x: u16) -> u16 {
    (x >> 8) | (x << 8)
}

fn should_skip_duplicate(symbols: &[(u32, String)], idx: usize, favor_last: bool) -> bool {
    if favor_last {
        // Skip if next symbol has the same offset
        if idx + 1 < symbols.len() && symbols[idx + 1].0 == symbols[idx].0 {
            return true;
        }
    } else {
        // Skip if previous symbol has the same offset
        if idx > 0 && symbols[idx - 1].0 == symbols[idx].0 {
            return true;
        }
    }
    false
}

/// Output file wrapper with tracking for append mode, base offset, etc.
struct OutputFile {
    writer: Box<dyn WriteSeek>,
    base_offset: usize,
    pos: usize,
}

impl OutputFile {
    fn current_offset(&self) -> usize {
        self.pos
    }
}

fn setup_output(
    file_name: &str,
    append_offset: i64,
    pointer_offset: u32,
    align_on_append: bool,
) -> Result<OutputFile> {
    if append_offset != 0 {
        let mut file = std::fs::OpenOptions::new()
            .read(true)
            .write(true)
            .open(file_name)?;

        let actual_append;
        if append_offset == -1 {
            // Append to end
            let end = file.seek(SeekFrom::End(0))? as usize;
            actual_append = if align_on_append && (end & 1) != 0 {
                file.write_all(&[0])?;
                end + 1
            } else {
                end
            };
        } else {
            actual_append = append_offset as usize;
            if align_on_append && (actual_append & 1) != 0 {
                eprintln!("Warning: An odd append offset is specified; the offset wasn't auto-aligned.");
            }
            file.seek(SeekFrom::Start(actual_append as u64))?;
        }

        // Write pointer if specified
        if pointer_offset != 0 {
            file.seek(SeekFrom::Start(pointer_offset as u64))?;
            file.write_all(&(actual_append as u32).to_be_bytes())?;
        }

        file.seek(SeekFrom::Start(actual_append as u64))?;

        Ok(OutputFile {
            writer: Box::new(file),
            base_offset: actual_append,
            pos: 0,
        })
    } else {
        let file: Box<dyn WriteSeek> = if file_name == "-" {
            Box::new(std::io::Cursor::new(Vec::new()))
        } else {
            Box::new(std::fs::File::create(file_name)?)
        };
        Ok(OutputFile {
            writer: file,
            base_offset: 0,
            pos: 0,
        })
    }
}
